test/wadler_test.rb

```diff
@@ -0,0 +1,136 @@
+# frozen_string_literal: true
+
+require 'prettyprint'
+
+require_relative 'lib'
+
+describe 'Wadler tests' do
+  it 'must work like ruby\'s PrettyPrint library' do
+    builder_block = proc { |margin, is_our_lib = true|
+      out = if is_our_lib
+        Oppen::Wadler.new(margin:)
+      else
+        PrettyPrint.new(''.dup, margin)
+      end
+      out.group {
+        out.group {
+          out.group {
+            out.group {
+              out.text 'hello'
+              out.breakable
+              out.text 'a'
+            }
+            out.breakable
+            out.text 'b'
+          }
+          out.breakable
+          out.text 'c'
+        }
+        out.breakable
+        out.text 'd'
+      }
+      if !is_our_lib
+        out.flush
+      end
+      out.output
+    }
+
```

Comment:
1. I'd rather extract the `builder_block` outside the `it`
1. Extract `test_lib_with_ref` outside the `it` and call it `assert_round_trip`
1. Write the tests in such a way where we loop over the tests data, generate the `it` for each assertion
1. Call the describe something a bit more meaningful, because we want all these tests to be rountrip equality assertions.

Reply:
The builder block varies depending on the it block, meaning that we might get too many blocks next to each other and cause confusion

Reply:
For (3.) I am not sure i undersand what you mean, do you want me to compare the output of PrettyPrint with Walder by looping from 1 to n and stop comparing to heredocs?